File: 46-permutations/46-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList();
        List<Integer> temp = new ArrayList();
        temp.add(nums[0]);
        ans.add(temp);
        for (int i = 1; i < nums.length; i++){
            List<List<Integer>> new_ans = new ArrayList();
            for (int j = 0; j <= i; j++){
                for (List<Integer> l: ans){
                    List<Integer> cur = new ArrayList(l);
                    cur.add(j, nums[i]);
                    new_ans.add(cur);
                }
            }
            ans = new_ans;
        }
        return ans;
    }
}

File: 46-permutations/46-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def findPerm(i, temp):
            if i >= len(nums) - 1:
                ans.append(temp[:])
                return
            
            for k in range(i, len(temp)):
                temp[i], temp[k] = temp[k], temp[i]
                findPerm(i + 1, temp)
                temp[i], temp[k] = temp[k], temp[i]
            
        ans = []
        findPerm(0, nums)
        return ans
        